src/systems/audio-zones-system.js

```diff
@@ -0,0 +1,161 @@
+  acc.coneOuterAngle = Math.min(acc.coneOuterAngle, curr.coneOuterAngle);
+  acc.coneOuterGain = Math.min(acc.coneOuterGain, curr.coneOuterGain);
+  return acc;
+}
+
+/**
+ * This system updates the audio-zone-sources audio-params based on the audio-zone-listener position.
+ * On every tick it computes the audio-zone-source and audio-zone-listener positions to check
+ * if the listener is inside/outside an audio-zone and applies the zone's audio parameters based.
+ * It only updates in case there has been any changes in the sources or listener's positions.
+ * If several audio-zones are in between the audio-zone-listener and the audio-zone-source, the applied
+ * audio parameters is a reduction of the audio-zones most restrictive audio parameters.
+ * i.e. If there are two audio-zones in between the listener and the source and the first one has gain == 0.1
+ * and the other has gain == 1.0, gain == 0.1 is applied to the source.
+ */
+export class AudioZonesSystem {
+  constructor(scene) {
+    this.scene = scene;
+    this.listener = null;
+    this.sources = [];
+    this.zones = [];
+    this.initialized = false;
+    this.ray = new THREE.Ray();
+    this.rayDir = new THREE.Vector3();
+    this.normalizedRayDir = new THREE.Vector3();
+    this.intersectTarget = new THREE.Vector3();
+  }
+
+  remove() {
+    this.sources = [];
+    this.zones = [];
+  }
+
+  setListener(listener) {
+    this.listener = listener;
+  }
+
+  unsetListener() {
+    this.listener = null;
+  }
+
+  registerSource(source) {
+    this.sources.push(source);
+  }
+
+  unregisterSource(source) {
+    const index = this.sources.indexOf(source);
+
+    if (index !== -1) {
+      this.sources.splice(index, 1);
+    }
+  }
+
+  registerZone(zone) {
+    this.zones.push(zone);
+  }
+
+  unregisterZone(zone) {
+    const index = this.zones.indexOf(zone);
+
+    if (index !== -1) {
+      this.zones.splice(index, 1);
+    }
+  }
+
+  tick() {
+    if (!this.scene.is("entered") || this.zones.length === 0) return;
+
+    // Update zones
+    this._updateZones();
+
+    this.sources.forEach(source => {
+      // Only check whenever either the source or the listener have updated zones (moved)
+      if (source.entity.isUpdated() || this.listener.entity.isUpdated()) {
+        // Cast a ray from the listener to the source
+        this.rayDir = source
+          .getPosition()
+          .clone()
+          .sub(this.listener.getPosition());
+        this.normalizedRayDir = this.rayDir.clone().normalize();
+        this.ray.set(this.listener.getPosition(), this.normalizedRayDir);
+
+        // First we check the zones the source is contained in and we check the inOut property
+        // to modify the sources audio params when the listener is outside the source's zones
+        // We always apply the outmost active zone audio params, the zone that's closest to the listener
+        const inOutParams = source.entity
+          .getZones()
+          .filter(zone => {
```

Comment:
I thought the same but this should only be called whenever there is a change in zones in a source or a listener so shouldn't happen very often so I  thought it might be ok.